#CP27 Setup-Continuous-Integration-Quality-Assurance-pipeline

feyton opened the pull request:
#### What does this PR do?

- [x] Add CI workflow with GitHub actions
- [x] Add build badge to the repository
- [x] Add Heroku pipeline
- [x] Add relevant badges to the repository read-me

### Description of Task to be completed
Configure a continuous integration pipeline using GitHub Actions and Heroku. The continuous integration service should automatically run tests and other preliminary build steps required to ship the next iteration of the product

### How should this be manually tested?
Check the Actions tab for the newly added action after creating this PR against the develop branch
Testing this by:
**Using the Github Command line tool (gh)**
**Note:** To install gh use **`winget install github.cli`** and use `cmd` to authenticate with the command `gh auth login`

```
gh pr list
gh pr checkout 6
npm i
touch ciTest.js
git add . && git commit -m "Testing message"
git push
gh browse
// head to actions and see the actions running the tests and deployment
```

### Any background context you want to provide?
Add the express package to serve the build folder on Heroku
Before merging the branch, the triggers will need to be changed to run against develop and the PR
### What are the relevant pivotal tracker stories?
#CP-27
### Screenshots (if appropriate)
![image](https://user-images.githubusercontent.com/28519395/153909221-814fa3f4-7c4d-404c-b2a7-1b91929e1bbf.png)

feyton committed:
ch-(add ci github actions)

Add the main yml and builds to the GitHub workflow.

feyton committed:
ch-(add ci github actions)

Add the main yml and builds to the GitHub workflow.

feyton committed:
ch-(add ci github actions)

Add the main yml and builds to the GitHub workflow.

feyton committed:
ch-(setup-ci) Add the heroku pipeline links to the readme

feyton committed:
ch-(setup-ci) Add the heroku pipeline links to the readme

feyton committed:
ch-(setup-ci) Add the heroku pipeline links to the readme

feyton committed:
Add the procfile for the Heroku pipeline

feyton committed:
Add the procfile for the Heroku pipeline

feyton committed:
Add the procfile for the Heroku pipeline

feyton committed:
Add the procfile for the Heroku pipeline

feyton committed:
--amend

feyton committed:
--amend

feyton committed:
--amend

feyton committed:
ch-(setup-pipeline)- add server to run on heroku

feyton committed:
ch-(setup-pipeline)- add server to run on heroku

feyton committed:
ch-(setup-pipeline)- add server to run on heroku

feyton committed:
ch-(setup-pipeline)- add server to run on heroku

feyton committed:
--amend

feyton committed:
--amend

feyton committed:
ch-(setup-ci)-testing the github actions
Merge branch 'ch-continuous-integration-cp-27' into ch-setup-ci-github-actions-cp-27

feyton committed:
ch-(setup-ci)-remove non-existent test script

feyton committed:
ch-(setup-ci)-remove non-existent test script

feyton committed:
ch-(setup-ci)-remove non-existent test script

feyton committed:
ch-(setup-ci)-remove non-existent test script

feyton committed:
Add the rporting workflow and secrets keys

feyton committed:
Add the rporting workflow and secrets keys

feyton committed:
Add the rporting workflow and secrets keys

feyton committed:
Add the rporting workflow and secrets keys

samyshyaka commented:
The implementation is good. I think we will have to wait for the Jest setup PR to be merged then we will be able to test this properly.

feyton committed:
Merge branch 'ch-setup-ci-github-actions-cp-27'
of https://github.com/atlp-rwanda/phantom-fe-codebandits into ch-setup-ci-github-actions-cp-27

feyton committed:
Add the bagdes and deployment workflow

With new addition of workflow specialized to deploy on
push on develop as well as the main branch after running the
tests successfully

feyton committed:
Add the bagdes and deployment workflow

With new addition of workflow specialized to deploy on
push on develop as well as the main branch after running the
tests successfully

feyton committed:
Fix the code climate CC_REPORT_ID typo

Add the coveralls repo token for coverage
publication.

feyton committed:
Fix the code climate CC_REPORT_ID typo

Add the coveralls repo token for coverage
publication.

feyton committed:
ch-(ci)- Fix the workflow run cancelling

Allow all runs to be completed

feyton committed:
ch-(ci)- Fix the workflow run cancelling

Allow all runs to be completed

feyton committed:
ch-(ci)- separation of concerns in ci

Refactored to leave jobs dedicated to same workflow

feyton committed:
ch-(ci)- separation of concerns in ci

Refactored to leave jobs dedicated to same workflow

feyton committed:
Fix mixed concurrency variables

The queed jobs cancels the ones before them
[Docs](https://github.community/t/concurrency-doesnt-work-as-expected/192205/6)

feyton committed:
Fix mixed concurrency variables

The queed jobs cancels the ones before them
[Docs](https://github.community/t/concurrency-doesnt-work-as-expected/192205/6)

feyton committed:
Add the working flow for pushing and pull reg

Against the develop branch. The deployment to the
staging app is only done on push/merge into the develop
while deployment to the main branch is done on push to main

feyton committed:
Merge branch 'ch-setup-ci-github-actions-cp-27' of https://github.com/atlp-rwanda/phantom-fe-codebandits into ch-setup-ci-github-actions-cp-27

feyton committed:
Add the favicon and meta tags to the web

samyshyaka commented on .codeclimate.yml:
```diff
@@ -0,0 +1,19 @@
+    enabled: true
+    config:
+      threshold: 50
+exclude_patterns:
```
Just trying to understand this. What does this line do exactly?  Are you excluding the tests and coverage?

feyton replied:
Yes, these settings tell code-climate the files to analyze. Since the tests are written extensively and can contain duplicate codes that can't be refactored, they are excluded. 
The coverage files are auto generated when we run npm coverage and are only used to send the reports to coveralls. Since they are JS files, they can raise issue of bad code while we have no control.

All files that are not user typed, should be excluded from code analysis and that is the purpose of the block

Jkadhuwa replied:
I think this file was done by @Abdulkeza, so Ideally it should not be appearing as modified in your PR. So let's refrain from modifying other people's codes without creating a bug or issues cards on Trello.

feyton replied:
### I will ensure that next time.

Although, I have seen such cases while I haven't edited a thing in a file. Not in this case though because I was sending the reports directly to the code-climate through GitHub actions I had the same file and resolved the conflict during re-base. 
Screenshot attached below, maybe that was the case. 
![image](https://user-images.githubusercontent.com/28519395/154038340-9ab3e674-42c1-4b91-bbde-f4db53bf12e4.png)

And so many others where we have App.js file edited while I only worked on the **workflow runs**. I am not able to pinpoint the why. But I guess it may have something to do with the **lint formatting** in some instances as well the _gitignore_ having file entry after they are staged.

samyshyaka commented on .github/workflows/heroku_deploy.yml:
```diff
@@ -0,0 +1,50 @@
+# This is a basic workflow to help you get started with Actions
```
There is no need to put comments on each line

feyton replied:
Since there are the basic markup for the workflow. That is why we have comments.
I will remove these

samyshyaka commented on .github/workflows/heroku_deploy.yml:
```diff
@@ -0,0 +1,50 @@
+    branches:
+      - develop
+  # Allows you to run this workflow manually from the Actions tab
+  workflow_dispatch:
```
There is nothing in the workflow_dispatch. I think it's better to remove this

feyton replied:
Actually not. The reason for that is for a user to call that action on demand. Without that empty entry it will only be triggered by the predefined actions like push

Yvad60 commented on .github/workflows/heroku_deploy.yml:
```diff
@@ -0,0 +1,50 @@
+      - uses: actions/checkout@v2
+      - uses: akhileshns/heroku-deploy@v3.12.12
+        with:
+          heroku_api_key: ${{secrets.HEROKU_API_KEY}}
```
how are we accessing this key? is that value a variable?

feyton replied:
That is a variable stored in GitHub repository secretes. Think of is a .env file which is not accessible to the public but accessible through a runner (Workflow action)

samyshyaka commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,53 @@
+    branches:
+      - main
+      - develop
+      - ch-setup-ci-github-actions-cp-27
```
Shouldn't the workflow be triggered only for the develop branch? I don't think it's necessary to include the other branches

feyton replied:
That line was added for testing the workflow. After the full review. I will send in a clean up commit for full configuration without unnecessary configurations.

samyshyaka reviewed (commented):
Nice work @feyton , Here below is some questions and feedback.

feyton committed:
Add the all branched as triggers

The entry for all branches is added to the main
remove unnecessary comments
